### PR TITLE
Wallet deployer and migrator for solc 0.5

### PR DESCRIPTION
**Pull request by i-stam**

### CODE REVIEW

File: test/internals/wallet-deployer/migrate_wallet_test.go [hunk 1]
```diff
@@ -0,0 +1,267 @@
+package wallet_deployer_test
+
+import (
+	// "context"
+	// "fmt"
+	// "math/big"
+	// "os"
+	// "testing"
```
Comment (draganm) [inline test/internals/wallet-deployer/migrate_wallet_test.go]: please remove comments

File: test/internals/wallet-deployer/migrate_wallet_test.go [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+
+	"github.com/ethereum/go-ethereum/common"
+	// "github.com/ethereum/go-ethereum/core/types"
```
Comment (draganm) [inline test/internals/wallet-deployer/migrate_wallet_test.go]: comments

File: test/internals/wallet-deployer/migrate_wallet_test.go [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+	When("no wallets are cached and a controller migrates a Wallet", func() {
+
+		BeforeEach(func() {
+			tx, err := WalletDeployer.MigrateWallet(Controller.TransactOpts(), Owner.Address(), RandomAccount.Address(), true, true, true, EthToWei(2), FinneyToWei(1), []common.Address{common.HexToAddress("0x1"), common.HexToAddress("0x2")})
+			Expect(err).ToNot(HaveOccurred())
+			Backend.Commit()
+			Expect(isSuccessful(tx)).To(BeTrue())
+		})
+
+		It("should return new Wallet address", func() {
```
Comment (draganm) [inline test/internals/wallet-deployer/migrate_wallet_test.go]: It's not quite clear what 'it' is here ... maybe `DeployedWallets(ownerAddress) should return then new Wallet address`

File: test/internals/wallet-deployer/migrate_wallet_test.go [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+		It("should emit MigratedWallet event", func() {
+
+			addr, err := WalletDeployer.DeployedWallets(nil, Owner.Address())
+			Expect(err).ToNot(HaveOccurred())
+
+			it, err := WalletDeployer.FilterMigratedWallet(nil)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(it.Next()).To(BeTrue())
+			Expect(it.Event.Wallet).To(Equal(addr))
+			Expect(it.Event.Owner).To(Equal(Owner.Address()))
```
Comment (draganm) [inline test/internals/wallet-deployer/migrate_wallet_test.go]: missing the check for `_oldWallet`

File: test/internals/wallet-deployer/migrate_wallet_test.go [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+				Expect(av.String()).To(Equal(EthToWei(1).String()))
+			})
+
+			It("should update the spend limit to 2 ETH", func() {
+				sl, err := MigratedWallet.SpendLimitValue(nil)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(sl.String()).To(Equal(EthToWei(2).String()))
+			})
+
+            It("should update the top up limit to 1 finney", func() {
```
Comment (draganm) [inline test/internals/wallet-deployer/migrate_wallet_test.go]: something went wrong with whitespace here ...

File: test/internals/wallet-deployer/migrate_wallet_test.go [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+			It("should emit MigratedWallet event", func() {
+
+				addr, err := WalletDeployer.DeployedWallets(nil, Owner.Address())
+				Expect(err).ToNot(HaveOccurred())
+
+				it, err := WalletDeployer.FilterMigratedWallet(nil)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(it.Next()).To(BeTrue())
+				Expect(it.Event.Wallet).To(Equal(addr))
+				Expect(it.Event.Owner).To(Equal(Owner.Address()))
```
Comment (draganm) [inline test/internals/wallet-deployer/migrate_wallet_test.go]: missing check for the old wallet address.

File: test/shared/backend_setup.go [hunk 7]
```diff
@@ -108,8 +111,8 @@ var CryptoFloatAddress common.Address
 var Licence *bindings.Licence
 var LicenceAddress common.Address
 
-var Wallet *bindings.Wallet
-var WalletAddress common.Address
+// var Wallet *bindings.Wallet
+// var WalletAddress common.Address
```
Comment (draganm) [inline test/shared/backend_setup.go]: please remove commented out code

File: test/wallet/wallet_suite_test.go [hunk 8]
```diff
@@ -33,13 +38,19 @@ func TestWalletSuite(t *testing.T) {
 var _ = BeforeEach(func() {
 	err := InitializeBackend()
 	Expect(err).ToNot(HaveOccurred())
+    // Deploy the Token wallet contract.
+    // var tx *types.Transaction
+	// WalletAddress, tx, Wallet, err = bindings.DeployWallet(BankAccount.TransactOpts(), Backend, Owner.Address(), true, ENSRegistryAddress, TokenWhitelistName, ControllerName, LicenceName, EthToWei(100))
+    // Expect(err).ToNot(HaveOccurred())
+	// Backend.Commit()
+	// Expect(isSuccessful(tx)).To(BeTrue())
```
Comment (draganm) [inline test/wallet/wallet_suite_test.go]: please remove the commented out code.

File: test/internals/wallet-deployer/migrate_wallet_test.go (hunk 1 above)
Comment (i-stam) [reply test/internals/wallet-deployer/migrate_wallet_test.go]: done

File: test/internals/wallet-deployer/migrate_wallet_test.go (hunk 4 above)
Comment (i-stam) [reply test/internals/wallet-deployer/migrate_wallet_test.go]: done

File: test/internals/wallet-deployer/migrate_wallet_test.go (hunk 5 above)
Comment (i-stam) [reply test/internals/wallet-deployer/migrate_wallet_test.go]: Github issue, looks fine on my editor

**Comment by mischat** [comment]
@i-stam and @draganm you guys should have a final look at this branch ... lots of rebasing

File: contracts/internals/walletDeployer.sol [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+    /// @notice parameters are passed in so that they can be used to construct new instances of the wallet
+    constructor(address _ens, uint _defaultSpendLimit) ENSResolvable(_ens) Controllable(controllerNode) public {
+        ens = _ens;
+        defaultSpendLimit = _defaultSpendLimit;
+    }
+
+    /// @dev Ether can be deposited from any source, so this contract must be payable by anyone.
+    function() external payable {
+        emit Received(msg.sender, msg.value);
```
Comment (draganm) [inline contracts/internals/walletDeployer.sol]: we should remove the event

File: contracts/internals/walletDeployer.sol [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+pragma solidity ^0.5.10;
+
+import "../wallet.sol";
+import "./controllable.sol";
+import "./ensResolvable.sol";
+
+//// @title Wallet deployer with pre-caching if wallets functionality.
+contract WalletDeployer is ENSResolvable, Controllable {
+
+    event Received(address _from, uint _value);
```
Comment (draganm) [inline contracts/internals/walletDeployer.sol]: don't think we need this event

**Pull request merged**